src/driver/Entry.ml

```diff
@@ -0,0 +1,146 @@
+  if String.equal !Typechecker.model_name "" then
+    Typechecker.model_name :=
+      mangle
+        (Flags.remove_dotstan
+           List.(hd_exn (rev (String.split model_name ~on:'/'))))
+      ^ "_model"
+  else Typechecker.model_name := mangle !Typechecker.model_name;
+  Typechecker.check_that_all_functions_have_definition :=
+    not flags.allow_undefined;
+  Transform_Mir.use_opencl := flags.use_opencl;
+  Lower_program.standalone_functions :=
+    flags.functions_only || flags.standalone_functions
+
+type other_output =
+  | Formatted of string
+  | DebugOutput of string
+  | Memory_patterns of string
+  | Info of string
+  | Version of string
+  | Generated of string
+  | Warnings of Warnings.t list
+
+type compilation_result = (string, Errors.t) result
+
+let stan2cpp model_name model (flags : Flags.t) (output : other_output -> unit)
+    : compilation_result =
+  let open Common.Let_syntax.Result in
+  reset_mutable_states model_name flags;
+  if flags.version then output (Info (Fmt.str "%s" version));
+  let ast, parser_warnings =
+    if flags.functions_only then
+      Parse.parse Parser.Incremental.functions_only model
+    else Parse.parse Parser.Incremental.program model in
+  output (Warnings parser_warnings);
+  let* result =
+    let* ast = ast in
+    if flags.debug_settings.debug_ast then
+      output (DebugOutput (fmt_sexp [%sexp (ast : Ast.untyped_program)]));
+    let+ typed_ast, type_warnings =
+      Typechecker.check_program ast
+      |> Result.map_error ~f:(fun e -> Errors.Semantic_error e) in
+    if flags.debug_settings.debug_typed_ast then
+      output (DebugOutput (fmt_sexp [%sexp (typed_ast : Ast.typed_program)]));
+    output (Warnings type_warnings);
+    if flags.info then output (Info (Info.info typed_ast));
+    let deprecation_warnings =
+      if flags.canonicalizer_settings.deprecations then []
+      else Deprecation_analysis.collect_warnings typed_ast in
+    output (Warnings deprecation_warnings);
+    if flags.auto_format then
+      output
+        (Formatted
+           (Pretty_print_prog.pretty_print_typed_program
+              ~bare_functions:flags.functions_only
+              ~line_length:flags.line_length
+              ~inline_includes:flags.canonicalizer_settings.inline_includes
+              ~strip_comments:flags.canonicalizer_settings.strip_comments
+              (Canonicalize.canonicalize_program typed_ast
+                 flags.canonicalizer_settings)));
+    let mir = Ast_to_Mir.trans_prog model_name typed_ast in
+    if flags.warn_uninitialized then
+      output (Warnings (Pedantic_analysis.warn_uninitialized mir));
+    if flags.warn_pedantic then
+      output (Warnings (Pedantic_analysis.warn_pedantic mir));
+    let () =
+      match flags.debug_settings.debug_mir with
+      | Off -> ()
+      | Basic ->
+          output (DebugOutput (fmt_sexp [%sexp (mir : Middle.Program.Typed.t)]))
+      | Pretty -> output (DebugOutput (Fmt.str "%a" Program.Typed.pp mir)) in
```

Comment:
This pattern occurs three times, first with `mir`, then with `tx_mir`, and then `opt_mir`. Let's make a helper function `debug_output mir`.